### PR TITLE
improved docker related things

### PR DESCRIPTION
**Pull request by maldex**
Following up issue #18, this pr concerns a few docker-things:
- Dockerfile: moved `add run.sh` to bottom, don't redo the whole thing if only this script has changed.
- run.sh: added an argument parser for classic Unix-compatibily and not only relying on environment variables.
- run.sh: overwrite the `nsx_image_webserver` in the config.yml with altered concourse_url.
- run.sh: corrected fly initialization as described in #18 (login/sync sequence and adding `--non-interactive`)
- nsx-t-install-09122018.tar: removed, afaik can the _releases​ feature_ of your RCS be used for this.
- all: some other little tweaks. Feel free to test and improve.

thx

### CODE REVIEW

**Comment by maldex** [comment]
Sorry, i don't get how that's supposed to work.

**Comment by kais66** [comment]
> Sorry, i don't get how that's supposed to work.

use '-s' when committing :)

File: docker_image/Dockerfile [hunk 1]
```diff
@@ -11,4 +10,6 @@ RUN apt-get update && \
     wget -O /usr/local/bin/fly https://github.com/Concourse/Concourse/releases/download/v3.14.1/fly_linux_amd64 && \
     chmod +x /usr/local/bin/fly
 
+ADD run.sh /home/run.sh
```
Comment (kais66) [inline docker_image/Dockerfile]: with this change, the layer for the RUN command will not be recreated?

File: docker_image/run.sh [hunk 2]
```diff
@@ -1,68 +1,117 @@
...
+        ;;
+esac
+done
+
+# defaults
+if [ -z "${CONCOURSE_TARGET}" ]; then		  CONCOURSE_TARGET=nsx-concourse; fi
+if [ -z "${PIPELINE_NAME}" ]; then 			  PIPELINE_NAME=nsx-t-install; fi
+if [ -z "${CONFIG_FILE_NAME}" ]; then         CONFIG_FILE_NAME=nsx_pipeline_config.yml; fi
+if [ -z "${NSXT_VERSION}" ]; then             NSXT_VERSION=2.3.0; fi
+if [ -z "${IMAGE_WEBSERVER_PORT}" ]; then     IMAGE_WEBSERVER_PORT=40002; fi
```
Comment (kais66) [inline docker_image/run.sh]: preferably this can default to 40001 for backward compatibility

File: docker_image/run.sh [hunk 3]
```diff
@@ -91,51 +140,63 @@ done
...
 
 # waiting for the concourse API server to start up
-while true; do
-	curl -s -o /dev/null $CONCOURSE_URL
-	if [[ $? -eq 0 ]]; then
-		break
-	fi
-	echo "waiting for Concourse web server to be running"
-	sleep 2
+while ! curl -s -o /dev/null $CONCOURSE_URL; do
```
Comment (kais66) [inline docker_image/run.sh]: 👍

File: docker_image/run.sh [hunk 4]
```diff
@@ -1,68 +1,117 @@
...
+fi 
 if [[ ! -e ${BIND_MOUNT_DIR}/${CONFIG_FILE_NAME} ]]; then
-	echo "Config file ${BIND_MOUNT_DIR}/${CONFIG_FILE_NAME} not found, exiting"
-	exit 1
+    echo "Config file ${BIND_MOUNT_DIR}/${CONFIG_FILE_NAME} not found, exiting"
+    exit 1
 fi
 
-# download the ovftool and OVA files
+log "strictly overwriting 'nsx_image_webserver'-value in ${CONFIG_FILE_NAME} - sorry for that"
```
Comment (kais66) [inline docker_image/run.sh]: seems we're eliminating the need for user to specify nsx_image_webserver. If that's the case, we can move it to pipeline_config_internal.yml. see line 105 to 108.

File: docker_image/run.sh [hunk 5]
```diff
@@ -91,51 +140,63 @@ done
...
+$fly_cmd_set
+
+log "fly: starting / unpausing pipeline"
+$fly_cmd_start
+
+log "fly: veryfly"
+fly targets ; fly -t $CONCOURSE_TARGET pipelines
+
+# adding aliases to .bashrc
+echo "alias fly-set=\"${fly_cmd_set}\"" >> ~/.bashrc
```
Comment (kais66) [inline docker_image/run.sh]: these updates around fly commands are great!
Can we still keep fly-reset and destroy ?

File: docker_image/Dockerfile (hunk 1 above)
Comment (maldex) [reply docker_image/Dockerfile]: Docker builds intermediate images for each directive, and the earlier a directive/subject has changed against a previous build, the more subsequent layers will be affected for a new build. The more things change, the later in Dockerfile| the more static things are, the earlier in Dockerfile.
So if i expect more regular changes in ubuntu than my run.sh script, i'd put the _apt-get_ later into the doker file.
For some applications, it's crucial you keep a certain sequence (e.g. install mysql before you run your create-tables.script.), in other cases it doesn't matter. My experience is like how you'd do imports on a class-file: System,Global,Project,Local,Namespace. Like you'd import _system.logger_ before _my.custom.logger_ in any language.   
Often it boils down to speed and correctness... Doing the big-jobs(apt-update, lots of MB from public mirror) before small-jobs (copy few KB from localdisk) helps both ...
Side-thing i started: i also add the Dockerfile itself for documentation. Did help me debug things a few times.

File: docker_image/run.sh (hunk 2 above)
Comment (maldex) [reply docker_image/run.sh]: Yes, sorry, changed port as part of my own testing. Just wanted to remove the _need_ for this to be set externally.

File: docker_image/run.sh (hunk 3 above)
Comment (maldex) [reply docker_image/run.sh]: as a spoiled _bash_-child, can someone tell me this would work in the _classic_ `/bin/sh` or `/bin/ksh`?

File: docker_image/run.sh (hunk 5 above)
Comment (maldex) [reply docker_image/run.sh]: you **should** keep whatever makes sense. As not mentioned: this PR is just a proposal, not a strict fix. You see the pattern, apply to it. I'm still learning about _fly_ so not much from my side. 

But, i guess i see what you're trying to achieve with the aliases, and personally i'd recommend to use shell-functions instead of aliases.
```
function list-files () { ls -lah $@; }
list-files /
```
As this specific container has a meta-purpose character [download ressources|start-file-server|compose-concourse|watch-workers/wait-for-termination??], most of what the run.sh script does could also be very well be provided against the interactive-mode `if [ -t 0 ] ; then ... else ... fi`[L182](https://github.com/maldex/nsx-t-datacenter-ci-pipelines/blob/master/docker_image/run.sh#L182). 

In a nutshell i'd:
- reduce the loads of variables being used, 
- group 'functionality' into functions, make them reusable (delete ova and downloaded stuff?)
- write all variables and functions (`declare -f list-files`) into `/nsx-t-install.rc`
- ... stay with the behaviour as --it or -d for now ...

**Comment by maldex** [comment]
Please dissect and reject this PR: it's merely how i quick-n-dirty got the nsx-t-pipeline up and displayable. It's notproduction safe but only addresses #18 and inherently a few things along. 
"Properly" addressing and implementing some of the discussed aspects would deserve another issue and pull-request afaik. Please invite me

File: docker_image/run.sh (hunk 5 above)
Comment (maldex) [reply docker_image/run.sh]: e.g.
```
function fly_cmd_set() {
    fly -t $CONCOURSE_TARGET set-pipeline -p $PIPELINE_NAME -c ${pipeline_dir}/pipelines/nsx-t-install.yml -l ${BIND_MOUNT_DIR}/${pipeline_internal_config} -l ${BIND_MOUNT_DIR}/${CONFIG_FILE_NAME} --non-interactive
}
function fly_cmd_start() {
    fly -t $CONCOURSE_TARGET unpause-pipeline -p $PIPELINE_NAME
}
function fly_cmd_stop() {
    fly -t $CONCOURSE_TARGET destroy-pipeline -p $PIPELINE_NAME --non-interactive
}
function fly_cmd_status() {
    fly targets ; fly -t $CONCOURSE_TARGET status ; fly -t $CONCOURSE_TARGET pipelines ; fly -t $CONCOURSE_TARGET workers
}

log "fly: setting the NSX-t install pipeline $PIPELINE_NAME"
fly_cmd_set

log "fly: starting / unpausing pipeline"
fly_cmd_start

log "fly: veryfly"
fly_cmd_status

declare -f fly_cmd_set fly_cmd_start fly_cmd_stop fly_cmd_status >> ~/.bashrc
export CONCOURSE_TARGET PIPELINE_NAME pipeline_dir BIND_MOUNT_DIR CONFIG_FILE_NAME pipeline_internal_config

source ~/.bashrc
bash 
```

File: docker_image/Dockerfile (hunk 1 above)
Comment (kais66) [reply docker_image/Dockerfile]: Thanks for the explanation. Good to know the best practice here. 
I think this change is of no risk and I'm also fine with adding the Dockerfile.

File: docker_image/run.sh (hunk 4 above)
Comment (kais66) [reply docker_image/run.sh]: It is a config param as the pipeline expects the information (i.e. the host IP, and port which nginx web server is listening on). However as you've already noticed, this information can be inferred from user input (the env vars set when running the docker container).

I didn't find a nice way to introduce default values to concourse pipeline variables. I ended up using an additional param file pipeline_config_internal.yml, created by the container, to be supplied to the pipeline. Because of the ordering of the param files specified when setting the pipeline:
1. vars in nsx_pipeline_config.yml takes precedence over the same vars specified in pipeline_config_internal.yml. 
2. also with pipeline_config_internal.yml, we can omit vars exposed to users for various reasons. 

I hope pipeline_config_internal.yml is a better place to put nsx_image_webserver

File: docker_image/run.sh (hunk 5 above)
Comment (kais66) [reply docker_image/run.sh]: Thanks for your suggestions. I agree shell functions are better than aliases and your snippet  looks good to me. I'm just trying to be cautious and to maintain the same experience for users when they switch to a new container. As such I would suggest separate the PR into two: 1. everything but the arg parser part; 2. arg parser; and we can get the 1st one in relatively quickly. I'll try to find time to do that but feel free to continue with the PR here.

Comment (maldex) [reply docker_image/run.sh]: will do, if i find time :)

File: docker_image/run.sh (hunk 4 above)
Comment (maldex) [reply docker_image/run.sh]: i cannot estimate that, still learning about concourse. But i like your approach, concat user and system config into one generated file. 
I'm again thinking about shell functions, like 'generate_pipeline', would keep the functionality for both application handling, classic shell and containerized. If i'll struggle into a nice solution i'll let you know :)
